Mrcs dev1

CCatania01 opened the pull request:
Added multiple connection CMRInet metrics and error data collector, CMRInetMetricsCollector
Created CMRInetMetrics class to display and monitor data
Added CMRInet metrics tests.
Added connection memo name to all window titles " - Connection <_memo>".
Changed Diagnostic menu name to Diagnostics
Fixed multiple connection Diagnostic bug. 
Created separate Diagnostic bundle, updated Bundle
Changed Diagnosics frame layout.  
  Added node address popup selection to restrict diags to configured nodes
  Test Setup panel shows options for selected test
  Added Inverted option checkbox to Test Setup panel
Added CTC switch heater plates, model board tower and station icons.

bobjacobsen commented on java/src/jmri/jmrix/cmri/serial/cmrinetmanager/CMRInetManagerFrame.java:
```diff
@@ -74,7 +74,7 @@ public void initComponents() {
 
         // set the frame's initial state
-        setTitle(Bundle.getMessage("WindowTitle"));
+        setTitle(Bundle.getMessage("WindowTitle") + " - Connection "+_memo.getUserName());
```
No need to change this, just a suggestion for later code, but you can make it a bit easier for the I18N crew by defining a new text string in the Bundle and have it handle the merge, similar to what you did elsewhere with the "StatusLine3" constant.

CCatania01 replied:
Will do this on the next release for all of the localization pieces.

bobjacobsen commented on java/src/jmri/jmrix/cmri/serial/diagnostic/DiagnosticBundle.properties:
```diff
@@ -0,0 +1,61 @@
+DiagnosticError11           = Error - In Card is not an Input Card in your Node definition.\nPlease try again.
+DiagnosticError12           = Error - In Card not 2 for SMINI, please try again.
+DiagnosticError13           = Error - Bad character in Filtering Delay field, please try again.
+StatusLine3                 = Port {0}, Bit {1} is ON - Compare LED's with the pattern below
```
The Bundle classes "search up", so when a string isn't found in a particular bundle, it's then searched in the parent (in this case, cmri.serial.Bundle), etc. That lets us move common phrases up, so that they don't have to be translated multiple times.  If there are redundant ones here, it would be could to delete them so they get picked up from above.  Not urgent, but the translators will thank you.

CCatania01 replied:
Thanks for the clarification on the search up.  I'll merge the added strings into the parent and remove the new bundle. This will be done in the next release.

bobjacobsen commented on java/test/jmri/jmrix/cmri/serial/cmrinetmetrics/CMRInetMetricsCollectorTest.java:
```diff
@@ -0,0 +1,44 @@
+    @Test
+    public void testCtor() {
+        Assume.assumeFalse(GraphicsEnvironment.isHeadless());
+        CMRInetMetricsCollectorTest action = new CMRInetMetricsCollectorTest(); 
```
Did you intend for this to create a new *Test object? Usually this test is creating the object under test itself, i..e CMRInetMetricsCollector.

CCatania01 replied:
What is missing are tests  CMRInetMetricsAction and CMRInetMetricsFrame under cmirnetmanager. I will create those.

bobjacobsen commented on java/test/jmri/jmrix/cmri/serial/cmrinetmetrics/CMRInetMetricsDataTest.java:
```diff
@@ -0,0 +1,37 @@
+    @Test
+    public void testMemoCtor() {
+        Assume.assumeFalse(GraphicsEnvironment.isHeadless());
+        CMRInetMetricsDataTest action = new CMRInetMetricsDataTest(); 
```
Is the right class being created here? (This might cause CI to fail with the CMRInetMetricsData showing no tests)

CCatania01 replied:
I am not sure what would be the appropriate tests for the MetricsCollector as it is started and runs headless, listening to the CMRInet traffic.  The MetricsData class is a data structure and access methods for the collected data.
I need some direction on how to approach the tests.

bobjacobsen commented:
Thanks!  Looks like a nice addition.

bobjacobsen commented:
CI SpotBugs found an issue:
```
[INFO] Dead store to dataButtonRenderer in jmri.jmrix.cmri.serial.cmrinetmanager.CMRInetMetricsFrame.initComponents() [jmri.jmrix.cmri.serial.cmrinetmanager.CMRInetMetricsFrame] At CMRInetMetricsFrame.java:[line 196] DLS_DEAD_LOCAL_STORE
```

Also there was a runtime failure. This one might be transient, so I've restarted it:
```
    [junit] ------------- ---------------- ---------------
2801    [junit] Testcase: testDispose(jmri.jmrix.cmri.serial.SerialTurnoutTest):	Caused an ERROR
2802    [junit] controller listeners remaining expected:<0> but was:<1>
2803    [junit] java.lang.AssertionError: controller listeners remaining expected:<0> but was:<1>
2804    [junit] 	at jmri.implementation.AbstractTurnoutTestBase.testDispose(AbstractTurnoutTestBase.java:88)
2805    [junit] 
2806    [junit] 
2807    [junit] Testcase: testDispose(jmri.jmrix.cmri.serial.SerialTurnoutTest):	Caused an ERROR
2808    [junit] controller listeners remaining expected:<0> but was:<1>
2809    [junit] java.lang.AssertionError: controller listeners remaining expected:<0> but was:<1>
2810    [junit] 	at jmri.implementation.AbstractTurnoutTestBase.testDispose(AbstractTurnoutTestBase.java:88)
2811    [junit] 
2812    [junit] 
```

bobjacobsen commented:
The `jmri.implementation.AbstractTurnoutTestBase.testDispose` error probably isn't transient, as it occurred in two Travis CI builds too.

bobjacobsen commented:
The `AbstractTurnoutTestBase.testDispose` error seems to be an overly fragile test. Working on it.

pabender commented:
@bobjacobsen, what leads you to believe the test is fragile?

The test works without issue for all the other turnout tests derived from it.

bobjacobsen commented:
The test is testing for "no listeners installed" on the TrafficController (scaffold) after `dispose()` is called on the Turnout under test.  That fails in this PR because something _else_ installed a listener as part of initialization, probably due to non-Turnout changes in this PR.  

What it should be testing for is something like "All listeners installed by the object under test have been removed".

CCatania01 commented:
Is the "no listeners installed" in the AbstractTurnoutTestBase error something I need to address?

bobjacobsen commented:
> +    @Test
> +    public void testCtor() {
> +        Assume.assumeFalse(GraphicsEnvironment.isHeadless());
> +        CMRInetMetricsCollectorTest action = new CMRInetMetricsCollectorTest(); 

I think I didn’t make my point clear. 

To test the crow of class Foo, the FooTest class should construct a Foo object, not a FooTest object.

bobjacobsen commented:
> Is the "no listeners installed" in the AbstractTurnoutTestBase error something I need to address?

No, I’m working on that.

bobjacobsen commented:
I've updated serial.SerialTurnoutTest to fix the test failure.

CCatania01 commented:
I corrected the CI error and added the connection title bundle with // NOI18N.
After commit, did push to upstream.  A dialog came up indicating changes needed to be pulled into local, allowed the operations from the dialog. Log shows Pull was cancelled.

PR shows no new build started.   Attached is the Netbeans log output.
[Push upstream with pulls.txt](https://github.com/JMRI/JMRI/files/1895080/Push.upstream.with.pulls.txt)

bobjacobsen commented:
The JUnit tests are passing now.

SpotBugs continues to report this error, which needs to be fixed:
```
[INFO] Dead store to dataButtonRenderer in jmri.jmrix.cmri.serial.cmrinetmanager.CMRInetMetricsFrame.initComponents() [jmri.jmrix.cmri.serial.cmrinetmanager.CMRInetMetricsFrame] At CMRInetMetricsFrame.java:[line 196] DLS_DEAD_LOCAL_STORE
```
But please note that the test coverage check hasn't been run yet due to the above, and might also find problems.

bobjacobsen commented:
> A dialog came up indicating changes needed to be pulled into local

Right. That's because I made the test changes.  You need to pull back from _your_ repository to fix this.  On the command line:
```
git pull https://github.com/CCatania01/JMRI.git
```

bobjacobsen commented:
Travis graphical restarted after what might be un-related:
```
[ERROR] Failures: 
[ERROR]   LayoutEditorTest.testSetShowHelpBar:399 getShowHelpBar
[ERROR]   OlcbSensorTest.testQueryState:246 expected:<[198f4c4c] 01 02 03 04 05 06 07 08> but was:<null>
[ERROR] Errors: 
[ERROR]   ExceptionDisplayFrameTest.testSetVisible:49->Object.wait:502->Object.wait:-2 Â» TestTimedOut
```

bobjacobsen commented:
Restart AppVeyor after it timed out past:
```
    [junit] WARN  - Found unknown Windows version: Windows Server 2012 R2 [Time-limited test] java.games.input.DefaultControllerEnvironment.getControllers()
```

bobjacobsen commented on java/test/jmri/jmrix/cmri/serial/cmrinetmetrics/CMRInetMetricsCollectorTest.java:
```diff
@@ -0,0 +1,44 @@
+    @Test
+    public void testStringCtor() {
+        Assume.assumeFalse(GraphicsEnvironment.isHeadless());
+        CMRInetMetricsCollectorTest action = new CMRInetMetricsCollectorTest(); 
```
This line needs to test the correct class.

bobjacobsen commented on java/test/jmri/jmrix/cmri/serial/cmrinetmetrics/CMRInetMetricsDataTest.java:
```diff
@@ -0,0 +1,37 @@
+    @Test
+    public void testMemoCtor() {
+        Assume.assumeFalse(GraphicsEnvironment.isHeadless());
+        CMRInetMetricsDataTest action = new CMRInetMetricsDataTest(); 
```
This line needs to test the correct class.

bobjacobsen commented:
A few Java warnings from the 1st Jenkins build:  http://jmri.tagadab.com/jenkins/job/Development/job/Builds/2455/warnings20Result/new/
